### ISSUE TITLE
Indexing fails to index all items with MSI enabled and a large gap in salable entity IDs

### ISSUE THREAD

**Issue by nthurston** [magento/magento2#34887]
### Summary
Related to #31224 which was closed because the issue didn't exist except with 3rd party module/plugin.  Same issue now exists with a built-in MSI plugin: module-inventory-catalog-search/Plugin/CatalogSearch/Model/Indexer/ChildProductFilterByInventoryStockPlugin.php

### Preconditions (*)

1. 2.4.3-p1
2. MSI Enabled

### Steps to reproduce (*)

1. Set display out of stock products to No
2. Create more than 2500 products
3. Create a new stock and source
4. Set website to use new stock
5. Assign all items to new source
6. Add inventory to source on first and last entity so that they are the only 2 in stock at that source
7. Re-index

### Expected result (*)

1. All products are indexed and appear in search/catalog/etc.

### Actual result (*)

1. All products after the gap are not indexed and are not visible on the site

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-Echo** [comment]
Hi @nthurston,

Thank you for reporting the issue.

However, we tried to reproduce it on Magento 2.4.3-p1/2.4-dev but the issue is not reproducible.
Steps followed to reproduce issue:

1. Installed the fresh Magento 2.4-dev/2.4.3-p1 instance.
2. Added the sample product data (40000 products).
3. Select all products except one and disable the rest and then reindex.
4. Observe the front end and it works fine, showed only one product as expected. (See the attached screenshots)

<img width="1428" alt="Screenshot 1" src="https://user-images.githubusercontent.com/51681435/147577535-c78d7ce7-4304-49f8-aa34-0f653ecde80a.png">
<img width="1428" alt="Screenshot 3" src="https://user-images.githubusercontent.com/51681435/147577741-e35a195f-5211-4d17-812d-eba100f2dc9b.png">
<img width="1428" alt="Screenshot 2" src="https://user-images.githubusercontent.com/51681435/147577561-eb60c38e-880d-4831-8fa7-f9d3bcbf5bce.png">

5. Again imported product, select all and perform product delete and then reindex.
6. Observe the result, all products are deleted (deleted around 23362 products). (See the attached screenshots)

<img width="1428" alt="Screenshot 4" src="https://user-images.githubusercontent.com/51681435/147577773-bdc15a37-d282-4fb1-a474-437ae374422c.png">
<img width="1428" alt="Screenshot 5" src="https://user-images.githubusercontent.com/51681435/147577777-eba75244-59c6-4e96-b728-4d15d5b9241d.png">

If the issue still persists at your end then please provide some more details like debug logs, screenshots or video, accordingly, I’ll try to reproduce it again.

Thank You

**Comment by nthurston** [comment]
That isn't exactly what I mentioned in the steps to reproduce.

Try leaving the first and the last entity id in either test.  So keep entity id 1 and entity id 40000, then disable or delete the rest, then re-index.

**Comment by nthurston** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Echo** [comment]
Hi @nthurston,

We tried the following scenario:
1. Imported 40000 products.
2. On the product page select all products except entity id 1 and entity id 40000.
3. Disable or delete the rest product and then re-index.
4. Verified the front end.

The frontend works fine, all of the other products are disabled/deleted only showing the reaming products(first and last).

See the attached screenshots:

<img width="1428" alt="Screenshot 1" src="https://user-images.githubusercontent.com/51681435/147675467-5e32368d-c910-4cab-a279-e8440148d12e.png">
<img width="1428" alt="Screenshot 2" src="https://user-images.githubusercontent.com/51681435/147675474-c18bc7db-91bb-4e1e-ab34-0c1cf6bcb297.png">
<img width="1428" alt="Screenshot 3" src="https://user-images.githubusercontent.com/51681435/147675483-d793c31b-5266-408a-93c1-9ebfd95478c7.png">
<img width="1428" alt="Screenshot 4" src="https://user-images.githubusercontent.com/51681435/147675486-1edb5241-ebad-4629-9d84-54fac38e19d8.png">
<img width="1428" alt="Screenshot 5" src="https://user-images.githubusercontent.com/51681435/147675489-77ec492c-2f78-4dd7-84e1-3f401b13ca56.png">

Can you please check and update if you did anything differently or advise if I missed anything ?

Thanks

**Comment by nthurston** [comment]
@magento give me 2.4.3-p1 instance

**Comment by nthurston** [comment]
Give me some time to reproduce on a development box, I don't think I will be able to reproduce on the github instance as I'm not able to reindex there.

Possible you need to have a custom stock as well.

**Comment by nthurston** [comment]
Ok - yes i reproduced.  I updated the original issue but the steps are as follows:
1. Set display out of stock products to No
1. Create more than 2500 products
1. Create a new stock and source
1. Set website to use new stock
1. Assign all items to new source
1. Add inventory to source on first and last entity so that they are the only 2 in stock at that source
1. Re-index

**Comment by engcom-Echo** [comment]
Hi @nthurston,

I have followed the steps that you have given and was able to reproduce the issue but it seems it's not an issue. 

After assigning a new source the new source status is out of stock and item quantity is zero which is causing the product to not display on the website which is expected.

After updating out of stock display to the “Yes” site works as expected. Please see the attached screenshots.
Also, I observed the system logs there is no issue logged in the system while performing these operations. 

<img width="1428" alt="Screenshot 2021-12-31 at 6 33 23 PM" src="https://user-images.githubusercontent.com/51681435/147825094-76780d34-c55a-464f-9a6c-6361ffee14d5.png">

<img width="1428" alt="Screenshot 2021-12-31 at 6 33 10 PM" src="https://user-images.githubusercontent.com/51681435/147825098-2c4e33c9-b9f4-49aa-9beb-5f715f00cd48.png">

Can you please try to follow the above changes and reindex the product ? You issue will probably be resolved. Let me know if you still face any problem.

Thanks

**Comment by nthurston** [comment]
I think you are misunderstanding, here are screenshots from my dev.

Website is set to use Test stock which include Test source.  Both items are in stock at Test.  After re-index, only entity id 1 is showing, not entity id 10000.

![image](https://user-images.githubusercontent.com/4097064/147939659-2a12b0fb-70c3-462d-baee-51649348c220.png)
![image](https://user-images.githubusercontent.com/4097064/147939743-182ad4ce-17b5-4dcf-a750-71d03513b96c.png)
![image](https://user-images.githubusercontent.com/4097064/147939834-588d9819-d3a5-4ca9-8c03-433c80c83566.png)
![image](https://user-images.githubusercontent.com/4097064/147939857-434e1558-3f82-49e5-827c-15c34a9f5fd7.png)

**Comment by engcom-Echo** [comment]
Hi @nthurston,

I have tried to reproduce this issue on the fresh installation (Magento 2.4-dev/2.4.3-p1) but I am not able to reproduce it by performing the above steps. Could you please update the exact steps that you are following so that I can try to reproduce the same on our instance ?

Thanks

**Comment by pwalus** [comment]
Hello @engcom-Echo @nthurston  I've also encountered this issue. I did some debugging and found out that: 

In the file vendor/magento/module-catalog-search/Model/Indexer/Fulltext/Action/Full.php:386

```php
         / ** 2 point **/
        $products = $this->dataProvider
            ->getSearchableProducts($storeId, $staticFields, $productIds, $lastProductId, $this->batchSize);

        / ** 1 point **/
        while (count($products) > 0) {
            $productsIds = array_column($products, 'entity_id');
            $relatedProducts = $this->getRelatedProducts($products);
            $productsIds = array_merge($productsIds, array_values($relatedProducts));

            $productsAttributes = $this->dataProvider->getProductAttributes($storeId, $productsIds, $dynamicFields);

            foreach ($products as $productData) {
                
                 / ** 3 point **/
                $lastProductId = $productData['entity_id'];

                $productIndex = [$productData['entity_id'] => $productsAttributes[$productData['entity_id']]];
                if (isset($relatedProducts[$productData['entity_id']])) {
                    $childProductsIndex = $this->getChildProductsIndex(
                        $productData['entity_id'],
                        $relatedProducts,
                        $productsAttributes
                    );
                    if (empty($childProductsIndex)) {
                        continue;
                    }
                    $productIndex = $productIndex + $childProductsIndex;
                }

                $index = $this->dataProvider->prepareProductIndex($productIndex, $productData, $storeId);
                yield $productData['entity_id'] => $index;
            }
            
            / ** 2 and 4 point **/
            $products = $this->dataProvider->getSearchableProducts($storeId, $staticFields, $productIds, $lastProductId, $this->batchSize);
        }
```

1. I want to reindex all products in catalogsearch so I run on a server `bin/magento index:reindex catalogsearch_fulltext`
2. Magento reindexes products until condition `while (count($products) > 0)` is not met.
2. In every iteration Magento gets 500 products (based on the $this->batchSize)
3. In every iteration Magento assigns the last product id to the variable $lastProductId
4. $lastProductId is then used to find next 500 products starting from $lastProductId

Now, because Magento has configured `Show out of stock == 0`, the plugin vendor/magento/module-inventory-catalog-search/Model/Indexer/FilterProductByStock.php:104 is being loaded. This plugin filters all products that are **out of stock** and then it returns array with products **in stock**. 

And for example as @nthurston tried to explain:
- We have 2500 products
- The first product is in stock
- The last one is in stock

With current logic, 
- Magento gets first 500 products and filters them with the plugin so in $products array there is only one item (the first one)
- Magento uses $lastProductId to find next 500 products and filters them with the plugin and since next 500 products are out of stock, $products array is empty so condition `while (count($products) > 0) ` is not met and stops reindexing rest of the products
- The last product which should be in stock is never reindexed...

I hope this explains the problem.

**Comment by hostep** [comment]
I have a suspicion that the changes in https://github.com/magento/inventory/commit/c06fcb17ba59f829bda496d0b46649996fe0ca63 might fix this, but it's a guess, I might be wrong because I didn't test anything.

**Comment by jordanvector** [comment]
@hostep that PR does actually indeed solve the issue for us at least and also this issue is easily reproducible if there is a gap larger than 500 between salable entity ids of the products, we had the same issue where indexers were coming back saying they completed successfully but we were only seeing 90 products in one of our stores than had 1000s of products in it.  After applying the changes https://github.com/magento/inventory/commit/c06fcb17ba59f829bda496d0b46649996fe0ca63 and reindexing there were 3673 products in the elasticsearch index

**Comment by nthurston** [comment]
@jordanvector Did you apply that whole commit or just the changes in Model/Indexer/FilterProductByStock.php and Plugin/CatalogSearch/Model/Indexer/ChildProductFilterByInventoryStockPlugin.php?

It seems like it's doing some other things and I'm struggling to parse if the rest of it is needed to fix this issue.

Need to fix on a production instance, but don't particularly want to run more untested code than needed.

**Comment by hostep** [comment]
@nthurston: I linked that particular commit, because it was the only merge commit I was able to find.
Unfortunately it consists of 3 different ticket (=fixes) instead of only one, so if you only need the fix for this gap problem, you'll probably want to combine all the separate [MC-42791](https://github.com/magento/inventory/search?q=MC-42791&type=commits) commits yourself.
Or do the opposite and filter out the changes from [MC-42002](https://github.com/magento/inventory/search?q=MC-42002&type=commits) and [MC-41545](https://github.com/magento/inventory/search?q=MC-41545&type=commits) (which might be faster/easier).

**Comment by jordanvector** [comment]
@nthurston correct on the changes made on the indexers matter to solve this issue at least in our case.

I made composer patches for this

```
"magento/module-inventory-catalog-search": {
    "Indexing fails to index all items with MSI enabled and a large gap in salable entity IDs #34887": "https://gist.githubusercontent.com/jordanvector/82614fd4a34b24041f915cecebcac89f/raw/4fea6c340149f0775d68a2462dedaaf2181f20ba/0001-fix-products-missing-from-search.patch"
 },
"magento/module-inventory-catalog-search-bundle-product": {
    "Indexing fails to index all items with MSI enabled and a large gap in salable entity IDs #34887": "https://gist.githubusercontent.com/jordanvector/cb7e6361f08c417898c0f8b2c1f2e270/raw/b38deff50dac05c17d7456ef32481a0d0fe1c101/0001-fix-inventory-search-indexer.patch"
},
"magento/module-inventory-catalog-search-configurable-product": {
    "Indexing fails to index all items with MSI enabled and a large gap in salable entity IDs #34887": "https://gist.githubusercontent.com/jordanvector/ef98f59d07880f445464512d303d3fc0/raw/9adb4d1c384ccf6326e8f96513cfe5e290dd01b2/0001-fix-inventory-search-indexer-config.patch"
},
```

**Comment by nthurston** [comment]
@jordanvector @hostep Thank you, appreciate it.  We ended up just futzing around with the entity ids a bit to solve for that client, but I'll probably try patching next time around.

**Comment by engcom-November** [comment]
Verified this issue with 2914 products on Magento 2.4-develop branch with MSI enabled but could not able to reproduce the issue.
Steps performed:
Stores - Configuration - Catalog - Inventory - Stock Options - Display out of stock products - Set to No
Catalog - Products - 2914 Products
Created New Source and stock - assigned main website and new source to new stock.
Reindex and clear cache
Catalog - Products - Select all Products - Assign inventory source - assigned new source and save
Catalog -  Products - Open 1st product and set Product to in stock and set quantity: 2 and save
Catalog - Products - Open last product and set Product to in stock and set quantity: 2 and save
Reindex and clear cache
Front-end - No issue: Both the products are displayed in catalog search after reindex
![image](https://user-images.githubusercontent.com/60198231/162742054-3764dd6d-a4f3-4106-91c6-23be475ffa12.png)

**Comment by jordanvector** [comment]
@engcom-November this test does not capture the true issue, in order to replicate we need a step in the middle that deletes at least 500 products to cause a gap in entity ids of 500 or more.

**Comment by engcom-November** [comment]
*Hi @jordanvector ,
I checked the issue again by deleting 550 Products to cause a gap in entity id's after step 7 but still cannot able to reproduce the issue. Kindly clarify us after which exact step, products needs to be deleted to reproduce the issue.
Thank you. 
**Steps performed:**
1. Stores - Configuration - Catalog - Inventory - Stock Options - Display out of stock products - Set to No
2. Catalog - Products - more than 2500 Products
3. Installed Inventory modules - Created New Source and stock - assigned main website and new source to new stock.
4. Reindex and clear cache
5. Catalog - Products - Select all Products - Assign inventory source - assigned new source and save
6. Catalog - Products - Open 1st product and set Product to in stock and set quantity: 2 and save
7. Catalog - Products - Open last product and set Product to in stock and set quantity: 2 and save
8. Delete 550 Products from Catalog - Products page (Ex: Deleted from page 6, 7 etc)
9. Reindex and clear cache
10. Front-end - No issue: Both the products are displayed in catalog search after reindex

**Comment by hostep** [comment]
I'm pretty sure that this has been solved in the Magento 2.4.4 release together with MSI version 1.2.4, so not sure if it's worth to keep testing this?

**Comment by nthurston** [comment]
Confirmed - original issue is resolved in 2.4.4.  I replicated the issue on a fresh 2.4.3-p1, then performed same steps on a fresh 2.4.4 and all appropriate items are indexed.

I'll go ahead and close.

**Comment by GamesmenJordan** [comment]
Assuming that there are two products in the category with the entity ids of 1 and 1000 for instance and a batch size of 500. If product 1 has no stock and product 1000 has stock, you would expect to see product 1000, instead you will see no products. If both products have stock, then both products will appear.

The issue is that the testing steps say both product should be in stock, but this is not correct.

The **batch before the gap** needs to meet these two criteria:

1. Has products in the batch
2. Has no products in stock

**Comment by hostep** [comment]
@GamesmenJordan: out of curiosity, do you still see this incorrect behavior on Magento 2.4.4 or higher? Because it's supposed to be fixed in 2.4.4 as far as I know.
